### PR TITLE
ICU-20685 Improving docs for UErrorCode.

### PR DESCRIPTION
**Pull request by sffc**
##### Checklist

- [x] Issue filed: https://unicode-org.atlassian.net/browse/ICU-20685
- [x] Updated PR title and link in previous line to include Issue number
- [x] Issue accepted
- [ ] Tests included
- [x] Documentation is changed or added

### CODE REVIEW

**Comment by sffc** [comment]
Ping @younies

**Comment by sffc** [comment]
Younies is OOO.  Reassigning to Markus.

File: icu4c/source/common/unicode/utypes.h [hunk 1]
```diff
@@ -385,17 +385,30 @@ typedef double UDate;
 /*===========================================================================*/
 
 /**
- * Error code to replace exception handling, so that the code is compatible with all C++ compilers,
- * and to use the same mechanism for C and C++.
+ * Standard ICU4C error code class, a substitute for exceptions.
```
Comment (markusicu) [inline icu4c/source/common/unicode/utypes.h]: s/class/type/

File: icu4c/source/common/unicode/utypes.h [hunk 2]
```diff
@@ -385,17 +385,30 @@ typedef double UDate;
...
+ *
+ * Initialize the UErrorCode with U_ZERO_ERROR, and check for success or
+ * failure using U_FAILURE() or U_SUCCESS():
```
Comment (markusicu) [inline icu4c/source/common/unicode/utypes.h]: confusing to have different orders of "success or failure" vs. "U_FAILURE() or U_SUCCESS()". Please swap the latter.

File: icu4c/source/common/unicode/utypes.h [hunk 3]
```diff
@@ -385,17 +385,30 @@ typedef double UDate;
...
+ *
+ *     UErrorCode errorCode = U_ZERO_ERROR;
```
Comment (markusicu) [inline icu4c/source/common/unicode/utypes.h]: Add `<pre>...</pre>` tag lines.

File: icu4c/source/common/unicode/utypes.h [hunk 4]
```diff
@@ -385,17 +385,30 @@ typedef double UDate;
...
+ *     // call ICU API that needs an error code parameter.
+ *     if (U_FAILURE(errorCode)) {
+ *         // An error occurred. Handle it here.
+ *     }
+ *
+ * C++ code should use icu::ErrorCode, available in errorcode.h.
```
Comment (markusicu) [inline icu4c/source/common/unicode/utypes.h]: add at the end: ", or a suitable subclass"

Comment (markusicu) [inline icu4c/source/common/unicode/utypes.h]: Prefix the header file name with unicode/ -- we want our users to use the prefix consistently.

File: icu4c/source/common/unicode/utypes.h [hunk 5]
```diff
@@ -385,17 +385,30 @@ typedef double UDate;
...
+ *
+ * For more information, see:
+ * http://icu-project.org/userguide/conventions
+ *
+ * Note: by convention, ICU functions that take a reference (C++) or a pointer
+ * (C) to a UErrorCode first test:
+ *
+ *     if(U_FAILURE(errorCode)) { return immediately; }
```
Comment (markusicu) [inline icu4c/source/common/unicode/utypes.h]: `pre` tags

File: icu4c/source/common/unicode/utypes.h [hunk 6]
```diff
@@ -385,17 +385,30 @@ typedef double UDate;
...
+ *     if (U_FAILURE(errorCode)) {
+ *         // An error occurred. Handle it here.
+ *     }
+ *
+ * C++ code should use icu::ErrorCode, available in errorcode.h.
+ *
+ * For more information, see:
+ * http://icu-project.org/userguide/conventions
+ *
+ * Note: by convention, ICU functions that take a reference (C++) or a pointer
```
Comment (markusicu) [inline icu4c/source/common/unicode/utypes.h]: s/by/By/ (sentence titlecase)

File: icu4c/source/common/unicode/utypes.h (hunk 3 above)
Comment (sffc) [reply icu4c/source/common/unicode/utypes.h]: Not necessary in doxygen markdown.

**Pull request merged**